misoclib/cpu/mor1kx.py

```diff
@@ -68,13 +131,112 @@ def __init__(self, platform, reset_pc):
-                                  i_dwbm_rty_i=0)
+                                  i_dwbm_rty_i=0,
+
+                                  i_du_addr_i=or1k_dbg_adr_i[:16],
+                                  i_du_stb_i=or1k_dbg_stb_i,
+                                  i_du_dat_i=or1k_dbg_dat_i,
+                                  i_du_we_i=or1k_dbg_we_i,
+                                  o_du_dat_o=or1k_dbg_dat_o,
+                                  o_du_ack_o=or1k_dbg_ack_o,
+                                  i_du_stall_i=or1k_dbg_stall_i,
+                                  o_du_stall_o=or1k_dbg_bp_o,
+                                  )
+
+        jtag_tap_tck = Signal()
+        jtag_tap_tdi = Signal()
+        jtag_tap_tdo = Signal()
+        jtag_tap_rst = Signal()
+        jtag_tap_idle = Signal()
+        jtag_tap_capture = Signal()
+        jtag_tap_shift = Signal()
+        jtag_tap_pause = Signal()
+        jtag_tap_update = Signal()
+        jtag_tap_select = Signal()
+
+        dbg_adr_o = Signal(32)
+        self.debug = dbg = wishbone.Interface()
+
+        self.specials += Instance(
+            "adbg_top",
+            i_cpu0_clk_i=clk,
+            o_cpu0_rst_o=or1k_dbg_rst,
+            o_cpu0_addr_o=or1k_dbg_adr_i,
+            o_cpu0_data_o=or1k_dbg_dat_i,
+            o_cpu0_stb_o=or1k_dbg_stb_i,
+            o_cpu0_we_o=or1k_dbg_we_i,
+            i_cpu0_data_i=or1k_dbg_dat_o,
+            i_cpu0_ack_i=or1k_dbg_ack_o,
+            o_cpu0_stall_o=or1k_dbg_stall_i,
+            i_cpu0_bp_i=or1k_dbg_bp_o,
+
+            # TAP interface
+            i_tck_i=jtag_tap_tck,
+            i_tdi_i=jtag_tap_tdi,
+            o_tdo_o=jtag_tap_tdo,
+            i_rst_i=jtag_tap_rst,
+            i_capture_dr_i=jtag_tap_capture,
+            i_shift_dr_i=jtag_tap_shift,
+            i_pause_dr_i=jtag_tap_pause,
+            i_update_dr_i=jtag_tap_update,
+            i_debug_select_i=jtag_tap_select,
+
+            # Wishbone debug master
+            i_wb_clk_i=clk,
+            i_wb_dat_i=dbg.dat_r,
+            i_wb_ack_i=dbg.ack,
+            i_wb_err_i=dbg.err,
+
+            o_wb_adr_o=dbg_adr_o,
+            o_wb_dat_o=dbg.dat_w,
+            o_wb_sel_o=dbg.sel,
+            o_wb_cyc_o=dbg.cyc,
+            o_wb_stb_o=dbg.stb,
+            o_wb_we_o=dbg.we,
+            o_wb_cti_o=dbg.cti,
+            o_wb_bte_o=dbg.bte,
+        )
+
+        self.specials += Instance(
+            "xilinx_internal_jtag",
+            o_tck_o=jtag_tap_tck,
+            i_debug_tdo_i=jtag_tap_tdo,
+            o_tdi_o=jtag_tap_tdi,
+            o_test_logic_reset_o=jtag_tap_rst,
+            o_run_test_idle_o=jtag_tap_idle,
+            o_shift_dr_o=jtag_tap_shift,
+            o_capture_dr_o=jtag_tap_capture,
+            o_pause_dr_o=jtag_tap_pause,
+            o_update_dr_o=jtag_tap_update,
+            o_debug_select_o=jtag_tap_select,
+        )
+
+
+        #bscan = self.submodules.bscan = BscanSpartan6()
```

Comment:
This should be used. But I would instantiate that outside of `MOR1KX` (and thus makre it architecture independent) and then hook it up.